### ISSUE TITLE
Getting error while calling calculateResolvedType() on scope in Java Parser

### ISSUE THREAD

**Issue by mukulbindal** [javaparser/javaparser#4319]
I have below classes in my project.

MyRequest.java:
```
package com.example.test;

import lombok.Data;

@Data
public class MyRequest {
	private String action;
	private String name;
	private String data;
}
```

MyService.java:
```
package com.example.test;

public class MyService {
	public void doStuff(MyRequest request) {
		
		switch(request.getAction()) {
		case "A":
			request.getName();
			break;
		}
	}
}
```

I am trying to find the full class name of all objects which is calling any method. Below is my code. For simplicity, I have provided the bare minimum replicable code.

**Comment by mukulbindal** [comment]
Even if this does not work or a bug, is there any alternative way of doing what I am trying to do here?

**Comment by jlerbsc** [comment]
JP's operation seems fine to me. To resolve the MyRequest type, under certain conditions, Jp escalates the search to the parent node (here SwitchEntry), which in turn tries to resolve the expression defined in the "request.getAction()" selector.

It cannot resolve this expression because the declaration of this method is generated by lombok. 

What you can do is perhaps add a specialization to the ReflectionTypeSolver class by overriding the filterName(String name) method to take into account the class generated by lombok. I've never done this but I think it should work.

**Comment by mukulbindal** [comment]
@jlerbsc Sorry I didn't understand much. What I tried just now is - I created a new class and extended the ReflectionTypeSolver:

```
class SpecialReflectionTypeSolver extends ReflectionTypeSolver {
    protected boolean filterName(String name) {
        return super.filterName(name) || name.startsWith("lombok");
    }
}
```

Is this what you meant? Sorry but I have no idea how this `filterName()` method works.

**Comment by jlerbsc** [comment]
Probably something like that

    class SpecialReflectionTypeSolver extends ReflectionTypeSolver {
    protected boolean filterName(String name) {
        return super.filterName(name) || name.startsWith("com.example.test");
    }
    }

Make sure that lombok actually generates java classes and that these classes are defined in the classpath of your test. 

Another solution would be to package these classes in a jar and configure it in the JarTypeSolver

**Comment by mukulbindal** [comment]
@jlerbsc I may be wrong here, but my code is able to give correct output for `request.getAction()` method call and is only failing for `request.getName()`. I want to understand the difference here. If standalone `request.getAction()` is recognized without any special symbol solver, why is it failing while resolving `request.getName()`.

PS: I tried above both approaches, still getting same error.

**Comment by jlerbsc** [comment]
This is curious because the getAction method does not exist in the source code being analysed.

**Comment by mukulbindal** [comment]
@jlerbsc Thank you so much for your quick reply. To add more, if I modify my source code like below:

```
package com.example.test;

public class MyService {
	public void doStuff(MyRequest request) {
		var value = request.getAction();
		switch(value) {
		case "A":
			request.getName();
			break;
		}
	}
}
```

I get the same error. But if I modify it like below, it works perfectly:

```
package com.example.test;

public class MyService {
	public void doStuff(MyRequest request) {
		String value = request.getAction();
		switch(value) {
		case "A":
			request.getName();
			break;
		}
	}
}
```

The only difference between both is, in first snippet, I am storing the value of `request.getAction()` in a variable initialized by `var` and in second snippet, I have provided `String`. Anyways I am not allowed to modify the source code. Now I am really curious to understand how `calculateResolvedType()` works.

**Comment by jlerbsc** [comment]
In the first case, to find the concrete type of the variable 'value' JP must solve the request.getAction() expression, whereas in the second case the type of the variable is taken from the declaration.

**Comment by mukulbindal** [comment]
Hi @jlerbsc . First of all thanks for all your quick suggestions. I overlooked the solution you proposed earlier:
```
Another solution would be to package these classes in a jar and configure it in the JarTypeSolver
```
There are two things to keep in mind. Only adding JarTypeSolver will not work. There is a specific order I need to follow - First goes all your third party jars, then your project jar and at last your source code. Since my original jar was a spring boot project, it didn't work when I included it in `JarTypeSolver`. Later when I checked the structure of the jar, I found that it is actually a spring boot project and there is another `.jar.original` file that represents my classes. Now I have renamed that file temporarily via script and I am not getting any error. Hence I am closing this issue now.

Thanks for all your support!!
